polars/polars-arrow/src/kernels/ewm/average.rs

```diff
@@ -1,270 +1,173 @@
+        }
+        opt_ewma = match opt_x {
+            None => opt_ewma,
+            Some(&x) => {
+                non_null_cnt += 1;
+
+                let prev = opt_ewma.unwrap().clone();
+
+                let value = if adjust {
+                    let numer = prev * denom * one_sub_alpha + x;
+                    denom = T::one() + one_sub_alpha * denom;
+                    numer / denom
+                } else {
+                    x * alpha + prev * one_sub_alpha
+                };
+                Some(value)
             }
-            None => prev,
         };
-        // Safety:
-        // we allocated vals.len()
-        unsafe { out.push_unchecked(prev) };
-    }
-
-    (leading_null_count, out)
-}
-
-// this is the non-adjusted variant in pandas
-pub fn ewma_inf_hist_no_nulls<T, I>(vals: I, alpha: T) -> Vec<T>
-where
-    T: Float + AddAssign + Debug,
-    I: IntoIterator<Item = T>,
-    I::IntoIter: TrustedLen,
-{
-    let mut iter = vals.into_iter();
-    let len = iter.size_hint().1.unwrap();
-    if len == 0 {
-        return vec![];
-    }
-
-    let mut out = Vec::with_capacity(len);
-    let first = iter.next().unwrap();
-    out.push(first);
-    let one_sub_alpha = T::one() - alpha;
-
-    let mut prev = out[0];
-    for val in iter {
-        let output_val = val * alpha + prev * one_sub_alpha;
-        prev = output_val;
-
-        // Safety:
-        // we allocated vals.len()
-        unsafe { out.push_unchecked(output_val) }
-    }
-
-    out
-}
-
-// this is the non-adjusted variant in pandas
-/// # Arguments
-///
-/// * `vals` - Iterator of optional values
-/// * `alpha` - Smoothing factor
-///
-/// Returns the a tuple with:
-/// * `leading_null_count` - the amount of nulls that must be set by the caller
-/// * `smoothed values` - The result of the ewma
-///
-pub fn ewma_inf_hists<T, I>(vals: I, alpha: T) -> (usize, Vec<T>)
-where
-    T: Float + AddAssign + Debug,
-    I: IntoIterator<Item = Option<T>>,
-    I::IntoIter: TrustedLen,
-{
-    let mut iter = vals.into_iter();
-    let len = iter.size_hint().1.unwrap();
-    if len == 0 {
-        return (0, vec![]);
-    }
-
-    let mut out = Vec::with_capacity(len);
-
-    let leading_null_count = set_first_none_null(&mut iter, &mut out);
-    let one_sub_alpha = T::one() - alpha;
-    let mut prev = out[out.len() - 1];
-
-    for opt_val in iter {
-        let output_val = match opt_val {
-            Some(val) => {
-                // Safety:
-                // we add first, so i - 1 always exits
-                let output = val * alpha + prev * one_sub_alpha;
-                prev = output;
-                prev
-            }
-            None => prev,
+        let push_me = match non_null_cnt < min_periods {
+            true => None,
+            false => opt_ewma,
         };
-
-        // Safety:
-        // we allocated vals.len()
-        unsafe { out.push_unchecked(output_val) }
+        results.push(push_me);
```

Comment:
Is there a significant performance benefit of using `push_unchecked` over `push`?

Reply:
There is as perf benefit, but you must ensure you have allocated enough. Otherwise we write to oob memory.

Reply:
If we use an iterator and `collect_trusted` we get the same without needing unsafe.

Reply:
> If we use an iterator and collect_trusted we get the same without needing unsafe.

That would be neat, but the loop maintains state, has `if` statements, etc. so I'm not seeing a way to implement it as an iterator. I pushed the `unsafe` version.

Reply:
An iterator can access a state from outside the iterator. That's no problem. In fact IIRC a `for loop` is actually syntactic sugar for an iterator.